Fix crop aligning for "Original image"

jenshannoschwalm opened the pull request:
We must avoid aligning if cropping to original image aspect as we would re-calculate width&height with wrong data thus resulting in square-sized crops. This problem was only evident
- with elder edits (likely before c0a8430e647a40f2325c2c6d8ec37b0b875f81b1)
- and using "original image" ratio

We now check for valid aligning calculations so making sure.

Fixes #19919

@TurboGit @pass712 would you check?

pass712 commented:
Will check soon

pass712 commented:
Doesn't work for me too. Same behaviour as described.

TurboGit commented:
On thing sure, 5.2.1 is broken as much as 5.3.0 as just expanding/collapsing the crop module changes the crop area.

TurboGit commented:
For 1: I cannot answer I do crop with image aspect ratio 99% of the time and otherwise square.
For 2: No I do not see issue after changing the crop aspect ratio and collapsing/expanding crop. But I see an issue if I only move the crop area.

TurboGit commented:
Reading the code the change in `modify_roi_out()` is frightening and I'm not sure this is good for the release. We really need to discuss that.

TurboGit commented:
Signal or Element. Whatever is easier for you.

TurboGit commented:
@jenshannoschwalm : Not sure you have my id on both. Maybe jitsi meet ?

Just click: https://meet.jit.si/dt

TurboGit commented:
@jenshannoschwalm : The link to the archive: https://drive.google.com/file/d/1p1Bv5c-rMTVokVrIVPJIhN6PyxTJDEB7/view?usp=sharing

It also contains screen shot of the rendering of the thumbs in 5.2.1, in master and after edit (only expand/collapse crop).

My observations, I'm pretty sure the cx & cy are ok as the cropped area when loaded into darkroom (master) was correct. So the solution discussed should possibly save us here :)

Let me know if you need something else to help you.

TurboGit commented:
> The old portrait example unfortunately has cx/cy at zero :-(

How is that possible? I mean when I open the picture on 5.2.1 my crop area is properly kept and the top/left corner is on the right position and not on 0x0. Given that, my expectation what that only the width/height were somehow broken otherwise it would not be possible for the crop to work at all. So I'm lost now.

TurboGit commented:
This is what I have on 5.2.1 when hovering on crop in history:

<img width="346" height="503" alt="crop-data-5 2 1" src="https://github.com/user-attachments/assets/3be8e0df-3308-4a11-a4a2-9b7cb0707bc6" />

No zero there.

jenshannoschwalm commented:
That for sure breaks fresh edits from master. It seems to be fine for 5.2.1 edits...

TurboGit commented:
But maybe we cannot do better for master edit. Users using later are prepared for that.

pass712 commented:
Here you go.... https://magentacloud.de/s/KsBcJLGm5tbqroX

jenshannoschwalm commented:
1. squashed all to be in one commit
2. added some logs for -d params if things are ok
3. log-always all *repair* activity
4. some formatting for less reading errors
5. it's close to impossible to construct all or severe bad crop history steps but from my tests chances are not so bad to get something close or even exact.

For all testers:
1. Any reports about similar issues with crop mode other than related to use of "original image" in 5.2.1 and later edited in master would be highly appreciated (although leading to instant depression). Please provide raw and xmp ...
2. You could test without any debug flags as "repairing" will always be logged always.
3. Please remember, if you test you should test landscape and portrait images with "original image" and "flipped original image".

pass712 commented:
Thanks @jenshannoschwalm and @TurboGit for your persistence in finding this nasty bug and your late night work! I keep my fingers crossed all works well and will give it an intense test.

zisoft commented on src/iop/crop.c:
```diff
@@ -176,28 +174,118 @@ int legacy_params(dt_iop_module_t *self,
+    {
+      const float pwd = self->dev->image_storage.p_width;
+      const float pht = self->dev->image_storage.p_height;
+      const gboolean safe = pwd > 4.0f && pht > 4.0f;
```
This line fires a GitGuardian incident "Generic Password".
Of couse a false positive due to the variable name `pwd`...
